realm-transformer/src/main/kotlin/io/realm/transformer/RealmOptionalAPITransformer.kt

```diff
@@ -0,0 +1,88 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.realm.transformer
+
+import com.android.build.api.transform.Format
+import com.android.build.api.transform.Context
+import com.android.build.api.transform.Transform
+import com.android.build.api.transform.QualifiedContent.ContentType
+import com.android.build.api.transform.QualifiedContent.Scope
+import com.android.build.api.transform.QualifiedContent.DefaultContentType
+import com.android.build.api.transform.TransformInput
+import com.android.build.api.transform.TransformOutputProvider
+import com.google.common.collect.ImmutableSet
+import io.realm.annotations.internal.OptionalAPI
+import org.slf4j.LoggerFactory
+import java.util.HashSet
+
+class RealmOptionalAPITransformer : Transform() {
+
+    private val logger = LoggerFactory.getLogger("realm-logger")
+
+    override fun getName(): String? = "RealmOptionalAPITransformer"
+
+    override fun isIncremental(): Boolean = false
+
+    override fun getInputTypes(): MutableSet<ContentType>? = ImmutableSet.of(DefaultContentType.CLASSES)
+
+    override fun getScopes(): MutableSet<Scope>? = ImmutableSet.of(Scope.EXTERNAL_LIBRARIES)
+
+    override fun getReferencedScopes(): MutableSet<Scope>? = ImmutableSet.of(Scope.PROJECT, Scope.PROJECT_LOCAL_DEPS,
+            Scope.SUB_PROJECTS, Scope.SUB_PROJECTS_LOCAL_DEPS, Scope.EXTERNAL_LIBRARIES)
+
+    override fun transform(context: Context?,
+                           inputs: MutableCollection<TransformInput>?,
+                           referencedInputs: MutableCollection<TransformInput>?,
+                           outputProvider: TransformOutputProvider?,
+                           isIncremental: Boolean) {
+
+        val classNames = HashSet<String>()
+        inputs!!.appendAllToClassNames(classNames)
+        val refClassNames = HashSet<String>()
+        referencedInputs!!.appendAllToClassNames(refClassNames)
+        val classPool = createClassPool()
+        inputs.appendAllToClassPool(classPool)
+
+        classNames.filter { it.startsWith("io.realm.") }.forEach {
+            classPool.get(it).declaredMethods.forEach {
+                val optionalAPIAnnotation = it.getAnnotation(OptionalAPI::class.java) as? OptionalAPI
+                val dependenciesList = optionalAPIAnnotation?.dependencies?.toList()
+
+                if (optionalAPIAnnotation == null) {
+                    logger.info("${it.declaringClass.name} ${it.name} doesn't have @OptionalAPI annotation.")
+                } else if (dependenciesList == null || dependenciesList.size == 0) {
+                    throw IllegalArgumentException("${it.name} doesn't have proper dependencies: " +
```

Comment:
Why throw the error if the dependency list is empty? Doesn't that just imply we should remove the method?